pkg/storegateway/limiter.go

```diff
@@ -0,0 +1,150 @@
+package storegateway
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/thanos-io/thanos/pkg/store"
+	"github.com/weaveworks/common/httpgrpc"
+
+	"github.com/cortexproject/cortex/pkg/storage/tsdb"
+	"github.com/cortexproject/cortex/pkg/util"
+	util_log "github.com/cortexproject/cortex/pkg/util/log"
+	"github.com/cortexproject/cortex/pkg/util/validation"
+)
+
+type limiter struct {
+	limiter *store.Limiter
+}
+
+func (c *limiter) Reserve(num uint64) error {
+	return c.ReserveWithType(num, 0)
+}
+
+func (c *limiter) ReserveWithType(num uint64, _ store.StoreDataType) error {
+	err := c.limiter.Reserve(num)
+	if err != nil {
+		return httpgrpc.Errorf(http.StatusUnprocessableEntity, err.Error())
```

Comment:
Should the composite limiter also return `httpgrpc.Errorf`?

Reply:
Good point

Reply:
Should the error code be a 500 or a 429?

Reply:
Hmm I thought it'll be tough to distinguish queries that will succeed with lower QPS vs queries that will always fail, so preferred to reject them all under the umbrella of "fetched too much data"

Reply:
made a change to throw 422 only if the request is asking more than user/instance max capacity. otherwise, it throws 429 (too many requests)